activesupport/lib/active_support/cache.rb

```diff
@@ -249,27 +253,79 @@ def retrieve_pool_options(options)
+      #   values for Zlib's <tt>"\x78"</tt> signature:
+      #
+      #     module MyCompressor
+      #       def self.deflate(dumped)
+      #         # compression logic... (make sure result does not start with "\x78"!)
+      #       end
+      #
+      #       def self.inflate(compressed)
+      #         if compressed.start_with?("\x78")
+      #           Zlib.inflate(compressed)
+      #         else
+      #           # decompression logic...
+      #         end
+      #       end
+      #     end
+      #
+      #     ActiveSupport::Cache.lookup_store(:redis_cache_store, compressor: MyCompressor)
+      #
       # [+:coder+]
-      #   Replaces the default serializer for cache entries. +coder+ must
-      #   respond to +dump+ and +load+. Using a custom coder disables automatic
-      #   compression.
+      #   The coder for serializing and (optionally) compressing cache entries.
+      #   Must respond to +dump+ and +load+.
+      #
+      #   The default coder composes the serializer and compressor, and includes
+      #   some performance optimizations. If you only need to override the
+      #   serializer or compressor, you should specify the +:serializer+ or
+      #   +:compressor+ options instead.
       #
-      #   Alternatively, you can specify <tt>coder: :message_pack</tt> to use a
-      #   preconfigured coder based on ActiveSupport::MessagePack that supports
-      #   automatic compression and includes a fallback mechanism to load old
-      #   cache entries from the default coder. However, this option requires
-      #   the +msgpack+ gem.
+      #   The +:coder+ option is mutally exclusive with the +:serializer+ and
+      #   +:compressor+ options. Specifying them together will raise an
+      #   +ArgumentError+.
       #
       # Any other specified options are treated as default options for the
       # relevant cache operations, such as #read, #write, and #fetch.
       def initialize(options = nil)
-        @options = options ? normalize_options(options) : {}
+        @options = options ? validate_options(normalize_options(options)) : {}
 
         @options[:compress] = true unless @options.key?(:compress)
         @options[:compress_threshold] ||= DEFAULT_COMPRESS_LIMIT
 
-        @coder = @options.delete(:coder) { default_coder } || :passthrough
-        @coder = Cache::SerializerWithFallback[@coder] if @coder.is_a?(Symbol)
+        @coder = @options.delete(:coder) do
+          legacy_serializer = Cache.format_version < 7.1 && !@options[:serializer]
+          serializer = @options.delete(:serializer) || default_serializer
+          serializer = Cache::SerializerWithFallback[serializer] if serializer.is_a?(Symbol)
+          compressor = @options.delete(:compressor) { Zlib }
+
+          Cache::Coder.new(serializer, compressor, legacy_serializer: legacy_serializer)
+        end
+
+        @coder ||= Cache::SerializerWithFallback[:passthrough]
```

Comment:
@jonathanhefner Hi there! 👋🏻 
Sorry to ping you on an already merged PR, but hopefully you'll see this! 🤞🏻

This change is breaking some stuff here at Shopify. We have some places that initialize a memory store this way:
```rb
ActiveSupport::Cache::MemoryStore.new(coder: :passthrough)
```

I understand I can fix our code by changing it to this:
```rb
ActiveSupport::Cache::MemoryStore.new(coder: ActiveSupport::Cache::SerializerWithFallback[:passthrough])
```

However, do you think this code could stay backwards compatible and allow setting a `coder` as a symbol?
One could for example add this line:
```rb
@coder = Cache::SerializerWithFallback[@coder] if @coder.is_a?(Symbol)
```
(Note this line existed as-is at LOC 272 before this PR.)

What do you think?

Reply:
@davidstosik we should fix our code. Passing a coder as a symbol make no sense.

I can help you if needed.

Reply:
Ok, so this wasn't supported in 7.0, this was added in 7.1 alpha and removed before release, so let's fix the Shopify side.

Reply:
> Is that something we officially support?

I think it would make sense. It's a fairly niche use case but can be useful with MemoryStore as a perf optimization if you know your values aren't mutable, or with MemcachedStore if you wish to use a serializer at the Dalli level.

Reply:
@casperisfine Thanks for digging that up! I already have a fix for our code so I think ignoring my message above is okay. 👍🏻